### PR TITLE
THRIFT-4452: optimize Dockerfile

### PR DESCRIPTION
**Pull request by RobberPhex**
Docker build flow:

1. add all repo
2. `apt-get update` (only one time)
3. install dependencies

### CODE REVIEW

File: build/docker/ubuntu-artful/Dockerfile [hunk 1]
```diff
@@ -41,10 +43,21 @@ RUN apt-get update && \
...
 # dotnet (core)
 RUN curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > /etc/apt/trusted.gpg.d/microsoft.gpg && \
-    echo "deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-ubuntu-artful-prod artful main" > /etc/apt/sources.list.d/dotnetdev.list
+    echo "deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-ubuntu-artful-prod artful main" > \
+    /etc/apt/sources.list.d/dotnetdev.list
 
-# node.js (this step runs apt-get update internally) - if we ever want a later version
-RUN curl -sL https://deb.nodesource.com/setup_8.x | bash -
+# node.js
+RUN curl -sL https://deb.nodesource.com/gpgkey/nodesource.gpg.key | sudo apt-key add - \
```
Comment (jeking3) [inline build/docker/ubuntu-artful/Dockerfile]: sudo is unnecessary and causes a failure in a docker container (see build)

File: build/docker/ubuntu-artful/Dockerfile [hunk 2]
```diff
@@ -41,10 +43,21 @@ RUN apt-get update && \
...
+    echo "deb https://deb.nodesource.com/node_8.x artful main" | sudo tee /etc/apt/sources.list.d/nodesource.list
+# D
+RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys EBCF975E5BA24D5E && \
+        wget http://master.dl.sourceforge.net/project/d-apt/files/d-apt.list -O /etc/apt/sources.list.d/d-apt.list && \
+        wget -qO - https://dlang.org/d-keyring.gpg | apt-key add -
+
+# Dart
+RUN curl https://dl-ssl.google.com/linux/linux_signing_key.pub | apt-key add - && \
+    curl https://storage.googleapis.com/download.dartlang.org/linux/debian/dart_stable.list > \
+    /etc/apt/sources.list.d/dart_stable.list
```
Comment (jeking3) [inline build/docker/ubuntu-artful/Dockerfile]: You need to run apt-get update here after adding these repositories.

File: build/docker/ubuntu-artful/Dockerfile [hunk 3]
```diff
@@ -41,10 +43,21 @@ RUN apt-get update && \
 
 # dotnet (core)
 RUN curl https://packages.microsoft.com/keys/microsoft.asc | gpg --dearmor > /etc/apt/trusted.gpg.d/microsoft.gpg && \
-    echo "deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-ubuntu-artful-prod artful main" > /etc/apt/sources.list.d/dotnetdev.list
+    echo "deb [arch=amd64] https://packages.microsoft.com/repos/microsoft-ubuntu-artful-prod artful main" > \
+    /etc/apt/sources.list.d/dotnetdev.list
```
Comment (jeking3) [inline build/docker/ubuntu-artful/Dockerfile]: Indent this so it makes more sense (it looks like a separate command)

File: build/docker/ubuntu-artful/Dockerfile (hunk 1 above)
Comment (jeking3) [inline build/docker/ubuntu-artful/Dockerfile]: where is the node.js setup run?

**Comment by jeking3** [comment]
I wish Dockerfile had a INCLUDE directive where common stuff could be factored out!

File: build/docker/ubuntu-artful/Dockerfile (hunk 2 above)
Comment (RobberPhex) [reply build/docker/ubuntu-artful/Dockerfile]: `apt-get install nodejs` run at line 137

**Pull request closed without merging**